### PR TITLE
Destroy NCCL communicator after every use 

### PR DESCRIPTION
**Pull request by keisukefukuda**
This PR fixes NCCL_UNHANDLED_CUDA_ERROR during CI. 

The recently-added mixed16 tests creates PureNcclCommunicator repeatedly in very short intervals. This may cause unhandled cuda error from NCCL library.

### CODE REVIEW

**Comment by kuenishi** [comment]
I would rather suggest

```python
if hasattr(communicator, 'nccl_comm'):
    communicator.nccl_comm.destroy()
```
than casually using `del` and `gc.collect()` in Python.

**Comment by shu65** [comment]
It is a good idea!
I confirmed that the unittest for chainermn were passed when using the following.
```
    if hasattr(communicator, 'nccl_comm'):
        communicator.nccl_comm.destroy()
    if hasattr(communicator, 'intra_nccl_comm'): # for hierarchical, single, and two_dimensional
        communicator.intra_nccl_comm.destroy()
```

File: tests/chainermn_tests/communicator_tests/test_communicator.py [hunk 1]
```diff
@@ -239,6 +239,20 @@ def create_communicator(param, use_gpu):
...
 
+def destroy_communicator(comm):
+    """Destroy internal NCCL communicator.
+
+    When too many NCCL communicator are alive, NCCL produces
+    unhandled CUDA error. To avoid this, we need to make sure to
+    destory NCCL communicator after every use."""
+    if hasattr(comm, 'nccl_comm') and comm.nccl_comm is not None:
+        comm.nccl_comm.destroy()
+    comm.nccl_comm = None
```
Comment (shu65) [inline tests/chainermn_tests/communicator_tests/test_communicator.py]: Is this line indent wrong?

Comment (keisukefukuda) [reply tests/chainermn_tests/communicator_tests/test_communicator.py]: Fixed! thanks!

**Comment by shu65** [comment]
Jenkins, test this please.

**Pull request merged**